src/api/hooks/useParachainEvents.ts

```diff
@@ -0,0 +1,71 @@
+import {useEffect, useState} from 'react';
+import {useLastBlock} from './useLastBlock';
+import type BN from 'bn.js';
+import {useApi} from 'context/ChainApiContext';
+import type {IEvent} from '@polkadot/types/types';
+import type {CandidateReceipt} from '@polkadot/types/interfaces';
+
+interface EventMapInfo {
+  blockHash?: string;
+  blockNumber?: BN;
+  relayParent: string;
+}
+
+type EventMap = Record<string, EventMapInfo>;
+
+interface LastEvents {
+  lastBacked: EventMap;
+  lastIncluded: EventMap;
+  lastTimeout: EventMap;
+}
+
+const EMPTY_EVENTS: LastEvents = {lastBacked: {}, lastIncluded: {}, lastTimeout: {}};
+
+function includeEntry(map: EventMap, event: Event, blockHash?: string, blockNumber?: BN): void {
+  const {descriptor} = (event as unknown as IEvent<[CandidateReceipt]>).data[0];
```

Comment:
`lastBlock?.events.forEach(({event, phase}) => {})` It is because the event we get from here is of type `Event` which is passed to the `includeEntry` function

Reply:
This is odd, is the `Event` a generic type?

Reply:
```
export interface Event extends GenericEvent {}
```

```
export declare class GenericEvent extends Struct implements IEvent<Codec[]> {
    constructor(registry: Registry, _value?: Uint8Array);
    /**
     * @description The wrapped [[EventData]]
     */
    get data(): GenericEventData;
    /**
     * @description The [[EventId]], identifying the raw event
     */
    get index(): EventId;
    /**
     * @description The [[EventMetadata]] with the documentation
     */
    get meta(): EventMetadataLatest;
    /**
     * @description The method string identifying the event
     */
    get method(): string;
    /**
     * @description The section string identifying the event
     */
    get section(): string;
    /**
     * @description The [[TypeDef]] for the event
     */
    get typeDef(): TypeDef[];
    /**
     * @description Converts the Object to to a human-friendly JSON, with additional fields, expansion and formatting of information
     */
    toHuman(isExpanded?: boolean): Record<string, AnyJson>;
}
```